src/charm.py

```diff
@@ -6,20 +6,355 @@
+
+        self.ingress_per_unit = IngressPerUnitProvider(self, self._ingress_relation_name)
+
+        # Charm events
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+
+        # Relation events
+        self.framework.observe(
+            self.ingress_per_unit.on.data_provided, self._on_ingress_data_provided
+        )
+
+    @property
+    def access_rules_configured(self) -> bool:
+        """Checks whether the access rules have been configured."""
+        return _is_not_empty(self.model.config.get("access_rules", ""))
+
+    def validate_access_rules_config(self, config: str) -> None:
+        """Validate that the access rules config is a valid json."""
+        try:
+            json.loads(config)
+        except json.JSONDecodeError as e:
+            raise InvalidConfigError(f"Failed to decode json: {e}")
+
+    @property
+    def _config(self) -> _RouteConfig:
+        return _RouteConfig(rule=self.config.get("rule"), root_url=self.config.get("root_url"))
+
+    @property
+    def rule(self) -> Optional[str]:
+        """The Traefik rule this charm is responsible for configuring."""
+        return self._config.rule
+
+    @property
+    def root_url(self) -> Optional[str]:
+        """The advertised url for the charm requesting ingress."""
+        return self._config.root_url
+
+    def _render_config(self, model_name: str, unit_name: str, app_name: str) -> RouteConfig:
+        return self._config.render(model_name=model_name, unit_name=unit_name, app_name=app_name)
+
+    @staticmethod
+    def _generate_traefik_unit_config(config: RouteConfig) -> UnitConfig:
+        rule, config_id, url = config.rule, config.id_, config.root_url
+
+        traefik_router_name = f"juju-{config_id}-router"
+        traefik_service_name = f"juju-{config_id}-service"
+
+        config: "UnitConfig" = {
+            "router": {
+                "rule": rule,
+                "service": traefik_service_name,
+                "entryPoints": ["web"],
+                "middlewares": ["oathkeeper"],
+            },
+            "router_name": traefik_router_name,
+            "service": {"loadBalancer": {"servers": [{"url": url}]}},
+            "service_name": traefik_service_name,
+        }
+        return config
+
+    @staticmethod
+    def _merge_traefik_configs(configs: Iterable["UnitConfig"]) -> "TraefikConfig":
+        traefik_config = {
+            "http": {
+                "routers": {config["router_name"]: config["router"] for config in configs},
+                "services": {config["service_name"]: config["service"] for config in configs},
+            }
+        }
+        return traefik_config
+
+    def _get_relation(self, endpoint: str) -> Optional[Relation]:
+        """Fetches the Relation for endpoint and checks that there's only 1."""
+        relations = self.model.relations.get(endpoint)
+        if not relations:
+            logger.info(f"no relations yet for {endpoint}")
+            return None
```

Comment:
`return` is good enough.

Reply:
When the return type is `Optional` you need to be explicit about the return (ie  `return None`), see https://github.com/python/mypy/issues/7511#issuecomment-689642560.